File: python/examples/math_worker_service.py
#!/usr/bin/env python3
"""
FILE: python/examples/math_worker_service.py
PURPOSE: Named v5 service example that registers as `math-service`.
"""

from __future__ import annotations

import math
import os

from multifrost import ServiceContext, ServiceWorker, run_service_sync


class MathService(ServiceWorker):
    def add(self, a: int, b: int) -> int:
        return a + b

    def multiply(self, a: int, b: int) -> int:
        return a * b

    async def factorial(self, n: int) -> int:
        if n < 0:
            raise ValueError("factorial is not defined for negative numbers")
        return math.factorial(n)


if __name__ == "__main__":
    peer_id = os.environ.get("MULTIFROST_PEER_ID", "math-service").strip() or "math-service"
    run_service_sync(MathService(), ServiceContext(peer_id=peer_id))

File: javascript/examples/math_worker_service.ts
/**
 * FILE: javascript/examples/math_worker_service.ts
 * PURPOSE: Show a minimal service peer built from ServiceWorker and runService.
 * OWNS: Service-side method dispatch example.
 * EXPORTS: None
 * DOCS: docs/spec.md, javascript/README.md
 */

import { runService, ServiceContext, ServiceWorker } from "../src/index.js";

class MathService extends ServiceWorker {
    async add(a: number, b: number): Promise<number> {
        return a + b;
    }

    async multiply(a: number, b: number): Promise<number> {
        return a * b;
    }

    async factorial(value: number): Promise<number> {
        let result = 1;
        for (let i = 2; i <= value; i += 1) {
            result *= i;
        }
        return result;
    }
}

void runService(
    new MathService(),
    new ServiceContext({
        peerId: "math-service",
    })
).catch(error => {
    console.error(error);
    process.exitCode = 1;
});
